### PR TITLE
Stable swap drink tests

### PR DESCRIPTION
**Pull request by JanKuczma**
Drink tests for `swap_exact_in(...)` method of the `StablePool` contract.

### CODE REVIEW

File: amm/drink-tests/src/stable_swap_tests.rs [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        [12, 18],
+        [100000000000000000, 100000000000000000000000],
+        1000,
+        6,
+        2000,
+        10000000000000000,
+        9999495232752197989995,
+    );
+}
+
```
Comment (deuszx) [inline amm/drink-tests/src/stable_swap_tests.rs:190]: Somewhere here we could put in a test with `0` amount being swapped – https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L782

**Review by deuszx — COMMENTED** [review]
Can we test sth with numbers that would overflow the calculations? Test extremes of percision/rate/decimals/etc.

File: amm/drink-tests/src/stable_swap_tests.rs (hunk 1 above)
Comment (deuszx) [reply amm/drink-tests/src/stable_swap_tests.rs:190]: There are 2 tests in Ref's repo that test with `amount_in=1` (https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L820-L854) where it verifies that the `out=0` (due to rounding I suppose) while the 1 token traded is still transferred to the pools' reserves.

**Review by deuszx — COMMENTED** [review]
There are also [these two](https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L858-L893) tests in the Ref that we're missing – can you tell me why you chose to ignore them? I can't see anything special about them at first glance.

**Comment by JanKuczma** [comment]
> Can we test sth with numbers that would overflow the calculations? Test extremes of percision/rate/decimals/etc.

This may be tricky since the overflow points depend on the number of tokens in the pool and parameter A. We can write tests for some specific configurations, e.g. 2-pool and MAX_A, 3-pool and MAX_A.


> There are also [these two](https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L858-L893) tests in the Ref that we're missing – can you tell me why you chose to ignore them? I can't see anything special about them at first glance.

I must have skipped them by mistake. Thanks!

File: amm/drink-tests/src/stable_swap_tests.rs (hunk 1 above)
Comment (JanKuczma) [reply amm/drink-tests/src/stable_swap_tests.rs:190]: > Somewhere here we could put in a test with `0` amount being swapped – https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L782

Yes we can, however it works slightly differently in our implementation: when `amount_in=0` swap function returns with an error `InsufficientInputAmount`.


> There are 2 tests in Ref's repo that test with `amount_in=1` (https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L820-L854) where it verifies that the `out=0` (due to rounding I suppose) while the 1 token traded is still transferred to the pools' reserves.

Yes, we should add similar tests but again, it works slightly differently than in our implementation so the tests should be accordingly adjusted.

**Review by deuszx — COMMENTED** [review]
There's a single test in [`rated_swap_tests.rs`](https://github.com/Cardinal-Cryptography/common-amm-stable-swap/blob/stable-rated-drink-tests/amm/drink-tests/src/rated_swap_tests.rs#L116) – can we bring it over to this file and delete the other?

File: amm/drink-tests/src/stable_swap_tests.rs [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        admin_fee_bps,
+    );
+    _ = stable_swap::add_liquidity(
+        session,
+        stable_swap,
+        BOB,
+        1,
+        initial_reserves.to_vec(),
+        bob(),
+    );
```
Comment (deuszx) [inline amm/drink-tests/src/stable_swap_tests.rs:102]: This returns `ContractResult` – let's `unwrap` etc etc just for the sake of consistency.

**Review by deuszx — APPROVED** [review]
Looks good, nice job 💪 

I imagine that we'll also want to add tests for `swap_exact_out` but if we add it to this file then it will grow to unreadable size. Maybe we can add a subfolder here `stable_swap` , add `mod.rs` there and split tests into file-per-endpoint? So we'd have `src/stable_swap/swap_exact_in.rs` `/drink-tests/src/stable_swap/swap_exact_out.rs` etc. Just an idea.

**Pull request merged**